core/src/main/scala/com/softwaremill/react/kafka_api_opt3/Example.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt3
+
+import scala.concurrent.Future
+
+import akka.Done
+import com.softwaremill.react.kafka_api_producer.Producer
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.apache.kafka.clients.producer.ProducerRecord
+
+trait Example {
+  implicit val ec = scala.concurrent.ExecutionContext.Implicits.global
+
+  def consumerProvider: () => KafkaConsumer[Array[Byte], String] = ???
+
+  def producerProvider: () => KafkaProducer[Array[Byte], String] = ???
+
+  class DB {
+    def save(offset: Long, data: String): Future[Done] = ???
+  }
+}
+
+// Consume messages and store a representation, including offset, in DB
+object Example3 extends Example {
+  val db = new DB
+
+  Consumer.source(consumerProvider)
```

Comment:
> But then I don't understand your original question. What should it do if you never commit? I mean, forget streams for now and just explain to me what you think are the desired semantics in normal non-stream terms so we can be on the same page (I fear some misunderstanding here).

I will try to clarify the situation. Are you talking about situation when my process which gets data from kafka is hanged and I want to shutdown? Sure there is no way except just force a shutdown and lost some data.

I was talking about case when your process can be shutdown in reasonable time. Sure I should be ready for hanging, but also I want to have correct predictable stream shutdown if everything goes ok to minimize data lost.